apis/apiextensions/fn/proto/v1beta1/run_function.proto

```diff
@@ -68,6 +68,24 @@ message RunFunctionRequest {
   // did not exist, Crossplane sets the map key to an empty Resources message to
   // indicate that it attempted to satisfy the request.
   map<string, Resources> extra_resources = 6;
+
+  // Optional credentials that this Function may use to communicate with an
+  // external system.
+  map <string, Credentials> credentials = 7;
+}
+
+// Credentials that a Function may use to communicate with an external system.
+message Credentials {
+  // Source of the credentials.
+  oneof source {
```

Comment:
Curious about the decision for `oneof` here. At first I thought it was for different sources, e.g. if something other than Secrets is supported in the future, but I think that's not being captured here, only the underlying raw data is stored here.

Is there another type of data that could be added here in the future you think? (i don't know protobufs well, could be a naive question 😇 )

Reply:
> At first I thought it was for different sources, e.g. if something other than Secrets is supported in the future

That's what it's for. The `oneOf` here in the `RunFunctionRequest` corresponds to the `source` differentiator field in the Composition. It's not a 1:1 mapping though.

> I think that's not being captured here, only the underlying raw data is stored here.

If Crossplane loads the credential data, the source isn't relevant to the function. For example `source: Secret` and a hypothetical `source: Vault` (not saying we should build that) would both just be `CredentialData` loaded and passed by Crossplane from the function's perspective.

Using a `oneOf` here lets us support credentials that can't be modeled as `CredentialData` in future. For example, what if we wanted to tell a function to use its pod identity (IRSA etc) as its credentials? That couldn't be modeled as `CredentialData`.

Reply:
Aha, that makes more sense now, thanks for takin the time to educate! 🙇‍♂️